modules/javafx.graphics/src/main/java/javafx/css/StyleConverter.java

```diff
@@ -86,10 +92,26 @@
  */
 public class StyleConverter<F, T> {
 
+    /**
+     * Indicates whether the style converter implements the {@link StyleConverter#convert(Map)}
+     * and {@link #convertBack(Object)} methods to enable object deconstruction and reconstruction.
+     * The following invariant must be satisfied: {@code convert(convertBack(value)).equals(value)}
+     *
+     * @since 24
+     */
+    @Documented
+    @Inherited
+    @Target(ElementType.TYPE)
+    @Retention(RetentionPolicy.RUNTIME)
+    public @interface SupportsReconstruction {}
+
+    final boolean supportsReconstruction;
```

Comment:
would it be easier just to pass a boolean to the constructor?
we are changing the public API anyway...

also: since the boolean is package protected, should we have a (short) javadoc for it?

Reply:
> would it be easier just to pass a boolean to the constructor? we are changing the public API anyway...

Usually boolean parameters are not a good API choice, but I have no strong opinion on this one. What do you prefer?

Reply:
I just think it's much simpler than an annotation...

Or it could be a public instance method `boolean isSupportReconstruction()` for the child classes to override, saving N (1? 8?) bytes per instance.

Reply:
But the annotation costs no bytes per instance, if that even matters, as I thought these were singletons so it looks like we're quite into micro-optimization territory here.

Reply:
Whether a style converter supports reconstruction should be a statement about the type, not a statement about the instance. For this reason, I don't prefer an instance method like `isSupportReconstruction()`. In any case, these are our options:
1. Boolean parameter / Feature enum passed to constructor
2. Instance method `isSupportReconstruction()`
3. Annotation
4. Empty marker interface
5. Interface with the `convertBack` method
6. Subclassing

I lean slightly towards the annotation, but I'm generally okay with either option.

Reply:
I just saw a boolean variable being instantiated from an annotation and thought "why jump through the multiple hoops?".  since there is a boolean, why not pass it directly?

it's less about memory allocation (though I would prefer to minimize that as well, but as @hjohn pointed out the difference is just a few bytes), but more about "entities must not be multiplied beyond necessity".

Reply:
Although I'm not advocating for a specific change, I think looking at how `T convert(Map)` works when it was newly introduced in Java 9 should be taken into account as well.

There seem to be a couple of general approaches:

1. Have a marker that can be checked from the outside
  - Annotation (checked from the outside)
    - Unusual for this kind of check
  - Marker interface
    - Seen more often, but unusual to not put the new method there
  - Subtype 
    - Bad idea, you can only inherit once
2. Introduce a new interface with the new method
  - Defines the method and serves as the marker at the same time
3. Have a method that can be called that guards a 2nd method call
  - Doesn't matter how this is fed (constructor, annotation, override)
    -  You see this more often, but it's not a great pattern (2 method calls for the price of one...)
4. Return a special value from a method that may be unsupported
  - The most obvious (modern) candidate is to return `Optional` but is a bit unusual to indicate support/non-support
  - Throw `UnsupportedOperationException` -- although standard, I think it indicates a programmer mistake that should be avoided (in other words, encountering this exception in production code should result in a code fix)
  - Return `null`

Now the very last option (returning `null`) was the way this was handled when `T convert(Map)` was introduced in Java 9.  Even though it may not be the prettiest solution, it has two things going for it:

- It's fast (probably the fastest of the above options, although for most by just a slim margin)
- It fits in with the existing design

So, my order of preference:
- Just return `null`
- A new interface, which includes the new method
- A marker only interface
- A supportsXYZ method (regardless of how that is approached)

I don't think I could get behind any of the other options, because they either stick out like a sore thumb versus the existing design, limit future options or just perform too poorly.

Reply:
Returning `null` seems fine from the perspective of `StyleConverter`, but it makes the calling code very awkward. Remember, we ended up here because we needed a way to detect whether an object would support component-wise transitions. If we can't detect that without invoking `convertBack`, we would need to either:
1. Speculatively decompose the value without knowing whether there even are any transitions defined on the node. This is bad because most of the time, there will be no transitions; we will end up deconstructing many objects for no reason.
2. _Assume_ that an object is component-transitionable, look up all potential transitions, and decompose the value; then, if we were wrong with out assumption, go back to the start and try again with another code path (`Interpolatable` or no transition).

Instead, what I've implemented now is a new interface `StyleConverter.WithReconstructionSupport`, which contains both methods:
```java
public interface WithReconstructionSupport<T> {
    T convert(Map<CssMetaData<? extends Styleable, ?>, Object> values);
    Map<CssMetaData<? extends Styleable, ?>, Object> convertBack(T value);
}
```

This allows us to use object deconstruction and reconstruction with a single interface reference, and also allows us to detect such support very easily.

Reply:
+1 for `StyleConverter.WithReconstructionSupport`

maybe use a shorter name, like

`StyleConverter.WithReconstruction`

Reply:
Clear, thanks MIchael.  I would suggest `Reconstructable` as the name, which seems more inline with interfacing naming (especially `Interpolatable`)

Reply:
But the converter is not reconstructable (in comparison to Interpolatable, which is something the marked types actually are).

Reply:
Sorry to keep harping on this, but so far, I think all the options are quite poor from the perspective of the user, and from a perspective of "could it have been designed this way in the first place". 

I also think I smell something fishy.  The `StyleConverter` seems to serve two unrelated purposes. One is to convert raw parsed values for the CSS parser (`convert(ParsedValue, Font)`) to a Java object.

The other is to support sub-properties (which you can also detect by checking if `CssMetaData.getSubProperties` is not empty).  Although that method is also called `convert` it is only used by  `CssStyleHelper`, which is completely unrelated to parsing.  A more accurate name for that method would be "consolidateSubPropertyValues" (and the opposite would then be something like "extractSubPropertyValues", or `consolidate` and `extract` for short).

There does not seem to be any overlap between these two purposes.  That is it to say, a style converter that is used for  decoding sub-properties is never used for parsing, and vice versa.  FX also does not support special syntax for specifying say a border in short hand form (like `-fx-border: dashed red`).

I think it may be a good idea to perhaps have a look if these purposes shouldn't just be split.  On the one hand, you have a style converter which uses `convert` for parsing.  On the other hand, you have `CssMetaData`, which when it has  sub-properties **must** support a back-and-forth translation (`convert` + `convertBack`) -- this is currently not enforced (ie. you can create `CssMetaData` with sub-properties but then **not** implement `convert(Map)` in the used style converter... this causes problems later on.

The only CSS property currently that supports both short-hand and sub-properties is actually the newly introduced `transition` property.  This could either implement two interfaces, or it could just be two classes.

I've done a quick attempt to do such a split (not looking too much at API) and the whole thing compiles and passes all tests still.  This is surprising, because I deleted several pieces of code, that are apparently never used (or perhaps not tested for, we can investigate this further).

See here: #1533

Diff: https://github.com/openjdk/jfx/pull/1533/commits/8f6d1e56a43068184599ad1ea47b4da98eb70bff#diff-fc96cf3204909a6110e64c37a935927be56efedc9f3f4f6f6843cc4fa7da5a1f

Let me know if you feel this has merit.

Reply:
I also think that we're probably dealing with two different things that are mixed up in `StyleConverter`. But instead of having this discussion here, it probably makes more sense to have it in the context of [JDK-8333121](https://bugs.openjdk.org/browse/JDK-8333121) because it seems relevant to eventually support both short-hand and long-hand property notations.

I've renamed `WithReconstructionSupport` to `SubPropertyConverter` and moved it to an internal package for now. Since both `BackgroundConverter` and `BorderConverter` are not public, we can change it any way we want at a later time.

Reply:
Yes, that's a good idea, I was mainly worried about introducing a new public API that we can't remove later when we haven't figured out yet what the best direction would be.  With this API removed, I think we can get this merged.

Reply:
One more thing.  You introduced `transition` short hand version.  Is it maybe a good idea to not do this yet until we have a better idea how to deal with them?  Shorthand properties are just a convenience; the feature will work without them.  If it was already released, then can ignore this as we're too late.

Reply:
The `transition` shorthand is already on track to be released, and we’re deep into RDP2.